dev/AppNotifications/AppNotificationBuilder/AppNotificationBuilder.cpp

```diff
@@ -0,0 +1,372 @@
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetHeroImage(winrt::Windows::Foundation::Uri const& imageUri)
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_heroImage.empty());
+
+        m_heroImage = L"<image placement=\"hero\" src=\"" + imageUri.ToString() + L"\"/>";
+
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetHeroImage(winrt::Windows::Foundation::Uri const& imageUri, winrt::hstring const& alternateText)
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_heroImage.empty());
+
+        m_heroImage = L"<image placement=\"hero\" src=\"" + imageUri.ToString() + L"\" alt=\"" + alternateText + L"\"/>";
+
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetAudioUri(winrt::Windows::Foundation::Uri const& audioUri)
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_audio.empty());
+
+        m_audio = L"<audio src=\"" + audioUri.ToString() + L"\"/>";
+
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetAudioUri(winrt::Windows::Foundation::Uri const& audioUri, AppNotificationDuration const& duration)
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_audio.empty());
+
+        m_audio = L"<audio src=\"" + audioUri.ToString() + L"\" loop=\"true\"/>";
+        m_duration = duration;
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetAudioEvent(AppNotificationSoundEvent const& soundEvent)
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_audio.empty());
+
+        // Need to map enums to strings :)
+        m_audio = L"<audio src=\"" + GetWinSoundEventString(soundEvent) + L"\"/>";
+
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetAudioEvent(AppNotificationSoundEvent const& soundEvent, AppNotificationDuration const& duration)
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_audio.empty());
+
+        // Need to map enums to strings :)
+        m_audio = L"<audio src=\"" + GetWinSoundEventString(soundEvent) + L"\" loop=\"true\"/>";
+        m_duration = duration;
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::MuteAudio()
+    {
+        THROW_HR_IF(E_INVALIDARG, !m_audio.empty());
+
+        m_audio = L"<audio silent=\"true\"/>";
+
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::AddButton(AppNotificationButton const& value)
+    {
+        THROW_HR_IF(E_INVALIDARG, m_inputList.size() == 5u);
+
+        m_inputList.push_back(value.as<winrt::Windows::Foundation::IStringable>().ToString());
```

Comment:
I don't think this is the desired behavior. Usually a builder waits until the last minute to lock things up. Capturing the xml for a button at the moment it is added instead of build time, prevents the caller to further modify the button.

Reply:
Should we throw here or the final AppNotification construction? This would apply to all cases @hulumane

Reply:
You should totally throw here. You're adding the button to the builder and we don;t support more than 5. You shouldn't get the string (build the button) until it's needed.

You'll have to keep a reference to the button in your input list, instead of the string already built.